src/python/pants/engine/rules.py

```diff
@@ -293,17 +290,39 @@ def rule_decorator(func, **kwargs) -> Callable:
     )(func)
 
 
-def validate_parameter_types(
-    func_id: str, parameter_types: Tuple[Type, ...], cacheable: bool
+def validate_requirements(
+    func_id: str,
+    parameter_types: Tuple[Type, ...],
+    awaitables: Tuple[AwaitableConstraints, ...],
+    cacheable: bool,
 ) -> None:
-    if cacheable:
-        for ty in parameter_types:
-            if issubclass(ty, SideEffecting):
-                # TODO: Technically this will also fire for an @_uncacheable_rule, but we don't
-                #  expose those as part of the API, so it's OK for this error not to mention them.
-                raise ValueError(
-                    f"A `@rule` that was not a @goal_rule ({func_id}) has a `SideEffecting` parameter: {ty}"
-                )
+    if not cacheable:
+        return
```

Comment:
Just realized now, that this early return cancels all the `Effect` checks further down for `@goal_rule`s. cc @stuhood

Reply:
Thanks! #14069.